File: es/es.js
const { Client } = require('@elastic/elasticsearch')
const logger = require('./logger.js');


const client = new Client({
    node: 'http://localhost:9200',
    auth: {
        username: 'elastic',
        password: 'password'
    },
  });

async function run_ping() {
    try {
        const response = await client.ping()
        if (response.statusCode === 200) {
            logger.info('Elasticsearch cluster is up!');
        } else {
            logger.error('Elasticsearch cluster is down!');
        };
    } catch (error) {
        logger.error('Error connecting to the Elasticsearch cluster:', error.message);
    }
}


const del_index = async (indexName) => {
    try {
        const del = await client.indices.delete({
            index: indexName,
        });
        logger.info('Index successfully deleted:');
        logger.info(body);
    } catch (error) {
        logger.error('Error deleting index:', error);
    }
};


const del_doc = async (indexName) => {
    try {
        const del = await client.delete({
            index: indexName,
            id: 'ypBAmIkBHSnjRtQvAkbH',
        });
        logger.info('Document successfully deleted:');
        logger.info(JSON.stringify(del));
    } catch (error) {
        logger.error('Error deleting document:', error);
    }
};


const update_doc = async (indexName) => {
    try {
        const updated = await client.update({
            index: indexName,
            id: '1',
            doc: {
                title: 'ai',
                news: 'new vibes',
            },
        });
        logger.info('Document updated successfully:');
        logger.info(JSON.stringify(updated));
    } catch (error) {
        logger.error('Error updating document:', error);
    }
};


const add_doc = async (indexName) => {
    try {
        const indexed = await client.index({
            index: indexName,
            id : '1',
            document: {
                title: 'ai',
                content: '1 hello world',
            }
        });
        logger.info('Document indexed successfully:');
        logger.info(JSON.stringify(indexed));
    } catch (error) {
        logger.error('Error indexing document:', error);
    }
};


const create_index = async (indexName) => {
    try {
        const idx = await client.indices.create({
            index: indexName,
            body: {
                mappings: {
                    properties: {
                        title: { type: 'text' },
                        content: { type: 'text' },
                    },
            },
            },
        });
        logger.info('Index successfully created:');
        logger.info(JSON.stringify(idx))
    } catch (error) {
        logger.error('Error creating index:', error);
    }
};


async function get_index(indexName) {
    try {
        const index_exists = await client.indices.exists({ index: indexName });
        if (!index_exists) {
            logger.info(`Index ${indexName} does not exist`);
            return;
        } else {
            logger.info(`Index ${indexName} exists`);
        }

        const obj = await client.indices.get({ 
            index: indexName 
        });
        logger.info('obj mappings:')
        logger.info(JSON.stringify(obj[indexName].mappings))

        const doc = await client.get({
            index: indexName,
            id: '1'
        });
        logger.info('doc:')
        logger.info(JSON.stringify(doc))
    } catch (error) {
        logger.error('Error getting index obj:', error);
    
    }
}


const search = async (indexName) => {
    try {
        const res = await client.search({
            index: indexName,
            query: {
                match_all: {},
            },
        });
        logger.info('Search Results:');
        if (res.hits && res.hits.hits) {
            const docs = res.hits.hits;
            docs.forEach((doc) => {
                logger.info(`doc: ${JSON.stringify(doc)}`);
            });
        } else {
            logger.info('No hits');
        }

        const resp = await client.search({
            query: {
                match: {
                    title: 'ai',
                },
            },
        });
        logger.info('get query results:')
        logger.info(JSON.stringify(resp))
    } catch (error) {
    logger.error('Error searching :', error);
    }

};


const start_app = () => {
    run_ping();
    // const indexName = 'new_index_js';
    const indexName = 'tweets'

    // del_index(indexName);
    // del_doc(indexName);

    // update_doc(indexName);
    // add_doc(indexName);
    // create_index(indexName);
    // get_index(indexName)



    search(indexName);
    
}

start_app();

File: es/python/es.py
from elasticsearch import Elasticsearch

# dict for each node
hosts = [
    {
        'host': 'localhost',
        'port': 9200,
        'scheme': 'http',  # Or 'https' if using SSL/TLS
    },
    ]
es = Elasticsearch(
    hosts=hosts,
    basic_auth=('elastic', 'password'),
    )


def delete_index():
    response = es.indices.delete(index='my_web_data_index')
    if response['acknowledged']:
        print("Index was deleted successfully")
    else:
        print("Index was not deleted successfully")


def update():
    document_id = '1'  
    script = {
        "source": "ctx._source.content = params.updated_value",
        "params": {
            "updated_value": "hi there updated"
        }
    }
    response = es.update(index=index_name, id=document_id, body={"script": script})
    print(response)


def delete_field():
    if 'field_name' in es.get(index=index_name, id='1')['_source']:
        update_script = {
            "script": {
                "source": "ctx._source.remove('field_name')"
            }
        }
        response = es.update(index=index_name, id='1', body=update_script)
        print(response)
    else:
        print("Field does not exist")


def mapping():
    mapping = {
        "properties": {
            "title": {"type": "text"},
            "message": {"type": "text"},
            "timestamp": {"type": "date"}
        }
    }
    print('Mapping created successfully')


def indexing():
    if not es.indices.exists(index=index_name):
        es.indices.create(index=index_name, body={"mappings": mapping})
    data_to_index = [
        {"title": "Doc1", "message": "Hi there", "timestamp": "2023-07-22"},
        {"title": "Doc2", "message": "hello world", "timestamp": "2023-07-22"},
    ]
    for idx, doc in enumerate(data_to_index):
        es.index(index=index_name, id=idx, body=doc)
    print("Indexed successfully")


def search():
    search_query = {
        "query": {
            "match_all": {}
        }
    }
    results = es.search(index=index_name, body=search_query)

    for hit in results["hits"]["hits"]:
        print(f"Document ID: {hit['_id']}")
        print(hit["_source"])
        # print(f"Score: {hit['_score']}, Title: {hit['_source']['title']}, Content: {hit['_source']['content']}")


if __name__ == '__main__':
    index_name = 'new_index'
    
    # delete_index()
    # update()
    # delete_field()
    # mapping()
    # indexing()
    search()